go/vt/grpcclient/client_auth_static.go

```diff
@@ -54,23 +62,56 @@ func (c *StaticAuthClientCreds) RequireTransportSecurity() bool {
 
 // AppendStaticAuth optionally appends static auth credentials if provided.
 func AppendStaticAuth(opts []grpc.DialOption) ([]grpc.DialOption, error) {
-	if credsFile == "" {
+	clientCredsLock.Lock()
+	defer clientCredsLock.Unlock()
+	if clientCredsErr != nil {
+		return nil, clientCredsErr
+	}
+	if clientCreds == nil {
 		return opts, nil
 	}
+	creds := grpc.WithPerRPCCredentials(clientCreds)
+	opts = append(opts, creds)
+	return opts, nil
+}
+
+// loadStaticAuthCredsFromFile loads static auth credentials from a file.
+func loadStaticAuthCredsFromFile(credsFile string) (*StaticAuthClientCreds, error) {
 	data, err := os.ReadFile(credsFile)
 	if err != nil {
 		return nil, err
 	}
-	clientCreds := &StaticAuthClientCreds{}
-	err = json.Unmarshal(data, clientCreds)
-	if err != nil {
-		return nil, err
-	}
-	creds := grpc.WithPerRPCCredentials(clientCreds)
-	opts = append(opts, creds)
-	return opts, nil
+	creds := &StaticAuthClientCreds{}
+	err = json.Unmarshal(data, creds)
+	return creds, err
 }
 
 func init() {
 	RegisterGRPCDialOptions(AppendStaticAuth)
+
+	clientCredsOnce.Do(func() {
+		if credsFile == "" {
+			return
+		}
+
+		clientCredsLock.Lock()
+		defer clientCredsLock.Unlock()
+		clientCreds, clientCredsErr = loadStaticAuthCredsFromFile(credsFile)
+		if clientCredsErr != nil {
+			return
+		}
+
+		sigChan := make(chan os.Signal, 1)
+		signal.Notify(sigChan, syscall.SIGHUP)
+		go func() {
+			for range sigChan {
+				if newClientCreds, err := loadStaticAuthCredsFromFile(credsFile); err == nil {
+					clientCredsLock.Lock()
+					clientCreds = newClientCreds
+					clientCredsErr = err
```

Comment:
I believe you want to always set this error var whether err is nil or not. Given that we check it in `AppendStaticAuth` which is what is being passed in as the grpc dial option.

Reply:
I wonder whether it is worth using the `lock` / `defer unlock` pattern at the beginning of this func. It means we will take and release the mutex even if the load func returns an error. On the other hand it is more future-proof to always `lock` and `defer unlock` at the beginning of a func. This func will be called infrequently and performance is not a concern.